# DEA CCR/CRS

In [1]:
import pandas as pd
import pyomo.environ as pyo

![tabela](tabela.png)

## Estruturação Matemática (Orientado a INPUTS):

    Max EffA: 85u1 + 120u2

    s.a:
    40v1 + 25v2
    85u1 + 120u2 - 40v1 - 25v2 <=0
    90u1 + 135u2 - 40v1 - 28v2 <=0
    92u1 + 160u2 - 40v1 - 30v2 <=0

    u1,u2,v1,v2 >=0

### Transferindo para o pyomo

In [2]:
modelo = pyo.ConcreteModel()

In [3]:
objetivo = [85,120,0,0]

In [4]:
restri = [[40,25,0,0],
          [85,120,-40,-25],
          [90,135,-40,-28],
          [92,160,-40,-30]]

In [5]:
ld = [1,0,0,0]

In [6]:
modelo.m = range(len(restri))
modelo.n = range(len(restri[0][:]))

In [7]:
modelo.n

range(0, 4)

In [8]:
modelo.x = pyo.Var(modelo.n, within=pyo.NonNegativeReals)

In [9]:
modelo.objetivo = pyo.Param(modelo.n, initialize=lambda modelo, j: objetivo[j])
modelo.restri = pyo.Param(modelo.m, modelo.n, initialize=lambda modelo, i, j: restri[i][j])
modelo.ld = pyo.Param(modelo.m, initialize=lambda modelo, i: ld[i])

In [10]:
def f_obj(modelo):
    return pyo.summation(modelo.objetivo, modelo.x)
modelo.z = pyo.Objective(rule=f_obj, sense=pyo.maximize)

In [11]:
def f_constr(modelo, i):
    return sum(modelo.restri[i,j] * modelo.x[j] for j in modelo.n) <= modelo.ld[i]

In [12]:
modelo.restrições = pyo.Constraint(modelo.m, rule=f_constr)

In [13]:
modelo.display()

Model unknown

  Variables:
    x : Size=4, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :  None : False :  True : NonNegativeReals
          1 :     0 :  None :  None : False :  True : NonNegativeReals
          2 :     0 :  None :  None : False :  True : NonNegativeReals
          3 :     0 :  None :  None : False :  True : NonNegativeReals

  Objectives:
    z : Size=1, Index=None, Active=True
ERROR: evaluating object as numeric value: x[0]
        (object: <class 'pyomo.core.base.var._GeneralVarData'>)
    No value for uninitialized NumericValue object x[0]
ERROR: evaluating object as numeric value: z
        (object: <class 'pyomo.core.base.objective.ScalarObjective'>)
    No value for uninitialized NumericValue object x[0]
        Key : Active : Value
        None :   None :  None

  Constraints:
    restrições : Size=4
        Key : Lower : Body : Upper
          0 :  None : None :   1.0
          1 :  None : None :   0.0

In [14]:
resultado = pyo.SolverFactory('gurobi', solver_io="python").solve(modelo)

In [15]:
l = list(modelo.x.keys())
for i in l:
    print("x" + str(i+1),'=', modelo.x[i]())

x1 = 0.0
x2 = 0.04
x3 = 0.16
x4 = 0.0


`Temos então, para que A seja Ótimo, que os pesos sejam:`

    u1 = 0.0
    u2 = 0.04
    v1 = 0.16
    v2 = 0.0

## Para o DMU-B

In [16]:
modelo2 = pyo.ConcreteModel()

objetivo2 = [90,135,0,0]

restri2 = [[40,28,0,0],
          [85,120,-40,-25],
          [90,135,-40,-28],
          [92,160,-40,-30]]

ld2 = [1,0,0,0]
modelo2.m = range(len(restri2))
modelo2.n = range(len(restri2[0][:]))

modelo2.x = pyo.Var(modelo2.n, within=pyo.NonNegativeReals)

modelo2.objetivo2 = pyo.Param(modelo2.n, initialize=lambda modelo2, j: objetivo2[j])
modelo2.restri2 = pyo.Param(modelo2.m, modelo2.n, initialize=lambda modelo2, i, j: restri2[i][j])
modelo2.ld2 = pyo.Param(modelo2.m, initialize=lambda modelo2, i: ld2[i])

def f_obj(modelo2):
    return pyo.summation(modelo2.objetivo2, modelo2.x)
modelo2.z = pyo.Objective(rule=f_obj, sense=pyo.maximize)

def f_constr(modelo2, i):
    return sum(modelo2.restri2[i,j] * modelo2.x[j] for j in modelo2.n) <= modelo2.ld2[i]

modelo2.restrições2 = pyo.Constraint(modelo2.m, rule=f_constr)

resultado = pyo.SolverFactory('gurobi', solver_io="python").solve(modelo2)

l = list(modelo2.x.keys())
for i in l:
    print("x" + str(i+1),'=', modelo2.x[i]())

x1 = 0.0
x2 = 0.03571428571428571
x3 = 0.14285714285714285
x4 = 0.0


`Temos então, para que B seja Ótimo, que os pesos sejam:`

    u1 = 0.0
    u2 = 0.03571428571428571
    v1 = 0.03571428571428571
    v2 = 0.0

## Para o DMU-C

In [18]:
modelo3 = pyo.ConcreteModel()

objetivo3 = [90,135,0,0]

restri3 = [[40,28,0,0],
          [85,120,-40,-25],
          [90,135,-40,-28],
          [92,160,-40,-30]]

ld3 = [1,0,0,0]
modelo3.m = range(len(restri3))
modelo3.n = range(len(restri3[0][:]))

modelo3.x = pyo.Var(modelo3.n, within=pyo.NonNegativeReals)

modelo3.objetivo3 = pyo.Param(modelo3.n, initialize=lambda modelo3, j: objetivo3[j])
modelo3.restri3 = pyo.Param(modelo3.m, modelo3.n, initialize=lambda modelo3, i, j: restri3[i][j])
modelo3.ld3 = pyo.Param(modelo3.m, initialize=lambda modelo3, i: ld3[i])

def f_obj(modelo3):
    return pyo.summation(modelo3.objetivo3, modelo3.x)
modelo3.z = pyo.Objective(rule=f_obj, sense=pyo.maximize)

def f_constr(modelo3, i):
    return sum(modelo3.restri3[i,j] * modelo3.x[j] for j in modelo3.n) <= modelo3.ld3[i]

modelo3.restrições3 = pyo.Constraint(modelo3.m, rule=f_constr)

resultado = pyo.SolverFactory('gurobi', solver_io="python").solve(modelo3)

l = list(modelo3.x.keys())
for i in l:
    print("x" + str(i+1),'=', modelo3.x[i]())

x1 = 0.0
x2 = 0.03571428571428571
x3 = 0.14285714285714285
x4 = 0.0


`Temos então, para que B seja Ótimo, que os pesos sejam:`

    u1 = 0.0
    u2 = 0.03571428571428571
    v1 = 0.14285714285714285
    v2 = 0.0